FinalTask/FinalTests.cs

```diff
@@ -0,0 +1,220 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net.Mail;
+using System.Runtime.InteropServices;
+using System.Text;
+using Allure.Commons;
+using Allure.Commons.Model;
+using Allure.NUnit.Attributes;
+using FinalTask.Pages;
+using Newtonsoft.Json.Linq;
+using NUnit.Framework;
+using NUnit.Framework.Interfaces;
+using OpenQA.Selenium;
+using OpenQA.Selenium.Chrome;
+using OpenQA.Selenium.Firefox;
+using OpenQA.Selenium.Remote;
+
+namespace FinalTask
+{
+    public enum RunConfiguration
+    {
+        Local,
+        Grid,
+        Cloud
+    }
+
+    [TestFixture(typeof(FirefoxDriver))]
+    [TestFixture(typeof(ChromeDriver))]
+    public class FinalTests<TWebDriver>: AllureReport where TWebDriver : IWebDriver, new()
+    {
+        IWebDriver driver;
+        RunConfiguration configuration = RunConfiguration.Local;
+        MainPage mainPage;
+
+        [OneTimeSetUp]
+        public void Init(){
+            switch (configuration)
+            {
+                case RunConfiguration.Local:
+                {
+                        driver = new TWebDriver();
+                }
+                break;
+                case RunConfiguration.Grid:
+                {
+                        if (typeof(TWebDriver) == typeof(ChromeDriver))
+                        {
+                            var capabilities = new ChromeOptions();
+                            driver = new RemoteWebDriver(new Uri("http://192.168.96.1:4445/wd/hub"), capabilities);
+                        } else {
+                            var capabilities = new FirefoxOptions();
+                            driver = new RemoteWebDriver(new Uri("http://192.168.96.1:4445/wd/hub"), capabilities);
+                        }
+                }
+                break;
+                case RunConfiguration.Cloud:
+                {
+                    if (typeof(TWebDriver) == typeof(ChromeDriver))
+                    {
+                        var capabilities = new ChromeOptions();
+                            driver = new RemoteWebDriver(new Uri("https://alexandraegorova:692364a6-3e80-434f-bc0a-748480390724@ondemand.eu-central-1.saucelabs.com:443/wd/hub"),
+                            capabilities);
+                        }
+                    else
+                    {
+                            var sauceOptions = new Dictionary<string, object>();
+                            var capabilities = new FirefoxOptions();
+                            capabilities.PlatformName = "Windows 10";
+                            capabilities.BrowserVersion = "76.0";
+                            capabilities.AddAdditionalCapability("sauce:options", sauceOptions, true);
+                            driver = new RemoteWebDriver(new Uri("https://alexandraegorova:692364a6-3e80-434f-bc0a-748480390724@ondemand.eu-central-1.saucelabs.com:443/wd/hub"), capabilities);
+                    }
+                }
+                break;
+            }
```

Comment:
This code will be better to store in a separate class (like DriverConfiguration) and then call necessary method from this class here in OneTimeSetUp